File: dotnet-mvc/CleanArchitecture/UseCases/ListProductsUseCase.cs
using CleanArchitecture.Entities;
using CleanArchitecture.UseCases.Ports;

namespace CleanArchitecture.UseCases;

/// <summary>
/// Use Case: List products with pagination.
/// This belongs to the USE CASES layer (application business rules).
/// Application rule: max page size is 100.
/// </summary>
public class ListProductsUseCase
{
    private readonly IProductGateway _gateway;

    public ListProductsUseCase(IProductGateway gateway)
    {
        _gateway = gateway;
    }

    public async Task<(IEnumerable<Product> Items, int TotalCount)> Execute(int page, int pageSize)
    {
        // Application rule: cap page size at 100
        if (pageSize > 100) pageSize = 100;
        if (pageSize < 1) pageSize = 10;
        if (page < 0) page = 0;

        return await _gateway.FindAll(page, pageSize);
    }
}

File: nodejs-mvc/clean-architecture/src/usecases/ListProductsUseCase.ts
import { Product } from '../entities/Product';
import { ProductGateway } from './interfaces/ProductGateway';
import { PaginatedOutput, ProductOutput } from './dto/ProductDTO';

/**
 * @layer Use Cases
 * @description Lists products with pagination.
 */
export class ListProductsUseCase {
  constructor(private readonly gateway: ProductGateway) {}

  async execute(page: number, size: number): Promise<PaginatedOutput<ProductOutput>> {
    const effectiveSize = Math.min(size, 100);
    const effectivePage = Math.max(page, 0);

    const { products, total } = await this.gateway.findAll(effectivePage, effectiveSize);

    return {
      content: products.map(this.toOutput),
      page: effectivePage,
      size: effectiveSize,
      totalElements: total,
      totalPages: Math.ceil(total / effectiveSize),
    };
  }

  private toOutput(product: Product): ProductOutput {
    return {
      id: product.id,
      name: product.name,
      description: product.description,
      price: product.price,
      category: product.category,
      stockQuantity: product.stockQuantity,
      sku: product.sku,
      active: product.active,
      createdAt: product.createdAt.toISOString(),
      updatedAt: product.updatedAt.toISOString(),
    };
  }
}
